### PR TITLE
Update Blob use to be compatible with Cypress 5.0

### LINKED ISSUE #214: File upload failures with new Cypress v5.0.0
cypress-file-upload plugin working correctly with Cypress v4.11.0 but after updating Cypress to v5.0.0 the file upload functionality fails.

Error: CypressError: `base64StringToBlob()` no longer returns a `Promise`. Update the use of `base64StringToBlob()` to expect a returned `Blob`

image

### PR DESCRIPTION
**Pull request by emilong**
Wraps return value of `Cypress.Blob.base64StringToBlob()` in a `Cypress.Promise.resolve`, which will allow us to treat it as a Promise no matter what. If it's already a Promise, as it was in Cypress <=4, this will be also be compatible. 

Fixes #214

### CODE REVIEW

**Review by slt — APPROVED** [review]

**Review by EtienneBruines — APPROVED** [review]

**Comment by mduft** [comment]
Any way I can test this? I tried:

- Clone & checkout of this branch
- npm run-script build
- npm pack
- On my project: npm install ../xxxx/cypress-file-upload-4.0.7.tgz
- Verified that the bundle.js contains the updated code
- Run cypress -> I still get the same error - why?

**Comment by EtienneBruines** [comment]
Cypress adds a `.then` function to the result of e.g. `base64StringToBlob`

Soure: https://github.com/cypress-io/cypress/blob/de175d910d02c8d082b05883885c8a7aa1fb27b7/packages/driver/src/util/breaking_change_warning.ts#L18
```js
        val.then = function () {
          $errUtil.throwErrByPath('breaking_change.blob_util2', {
            args: {
              functionName: key,
            },
          })
        }
```

Perhaps wrapping it around `Cypress.Promise.resolve(...)` still invokes the `.then` method that was added by Cypress? 

According to the bluebird docs:

> http://bluebirdjs.com/docs/api/promise.resolve.html
> If value is a thenable (Promise-like object, like those returned by jQuery's $.ajax), returns a trusted Promise that assimilates the state of the thenable.

**Comment by mduft** [comment]
Maybe @jennifer-shehane can shed some light on this? What else would have to be done to have a working blob-util again?

**Comment by chrisbreiding** [comment]
We added the `.then` to try and add a better error message when using the old API, but now it's clear that wasn't a great idea since it gets confused for a 'thenable'. 

You should be able to work around it by using `Promise.try` instead of `Promise.resolve`:

```javascript
// before:
Cypress.Promise.resolve(Cypress.Blob.base64StringToBlob(fileContent, mimeType))
// after:
Cypress.Promise.try(() => Cypress.Blob.base64StringToBlob(fileContent, mimeType)))
```

**Comment by mduft** [comment]
> ```js
> Cypress.Promise.try(() => Cypress.Blob.base64StringToBlob(fileContent, mimeType)))
> ```

I tried with this code instead, but I get the very same result... 

```
base64StringToBlob() no longer returns a Promise. Update the use of base64StringToBlob() to expect a returned Blob.
```

File: lib/file/getFileBlobAsync.js
Comment (chrisbreiding) [inline lib/file/getFileBlobAsync.js:10]:
```suggestion
    Cypress.Promise.try(() => Cypress.Blob.base64StringToBlob(fileContent, mimeType)),
  [ENCODING.BINARY]: (fileContent, mimeType) =>
    Cypress.Promise.try(() => Cypress.Blob.binaryStringToBlob(fileContent, mimeType)),
```

Adding the `() =>` is necessary for this to work.

Comment (mduft) [reply lib/file/getFileBlobAsync.js]: Yeah, `() =>` missing - however the code still does not work... :(

**Comment by emilong** [comment]
Thanks, @chrisbreiding! I missed that detail... still a bit early in the morning :sweat_smile: :coffee:

**Comment by EtienneBruines** [comment]
`Cypress.Promise.try` does not mean "meh, I don't care if it works" though.

It's just a fancy way of handling the non-async part of the potential errors.

> http://bluebirdjs.com/docs/api/promise.try.html
> Start the chain of promises with Promise.try. Any synchronous exceptions will be turned into rejections on the returned promise.

**Comment by chrisbreiding** [comment]
That's my bad. I suggested it before trying it out. I had hoped it wouldn't trigger using `.then` on the return value like `Promise.resolve`, but it looks like it does.

Might have to resort to an ugly hack to fix this. Something like the following:

```javascript
const removeThen = (value) => {
  if (value) value.then = undefined

  return value
}

Cypress.Promise.resolve(removeThen(Cypress.Blob.base64StringToBlob(fileContent, mimeType)))
```

**Comment by emilong** [comment]
I'm thinking we'll want to condition that on the Cypress version too? To be compatible with < 5.0.0. I'll draft it up.

**Comment by chrisbreiding** [comment]
Sorry for the hassle with this one. We'll improve this for the next release, but unfortunately the hack will have to remain for users on 5.0.0.

**Comment by emilong** [comment]
Ok, I actually did due diligence this time and confirmed it works on my tests locally for 4.12.1 and 5.0.0. :crossed_fingers:

**Comment by mduft** [comment]
Yeah, I can confirm that this fix work perfectly fine now :)

**Review by mduft — APPROVED** [review]

**Comment by mduft** [comment]
MaintainerNotAvailableException?

**Review by gregberge — APPROVED** [review]

**Comment by dragonfire1119** [comment]
Hope this gets merged soon this works perfectly!

**Comment by KatieMFritz** [comment]
Can anyone help me figure out how to use the https://github.com/binti-family/cypress-file-upload/tree/cypress-v5-compat branch in my build?

What I did:
1. Delete `cypress-file-upload` from node_modules directory.
2. Delete `yarn.lock`.
3. Add to `devDependencies` in `package.json`: `"cypress-file-upload": "binti-family/cypress-file-upload#cypress-v5-compat",
`
4. `yarn install`
5. Add to `cypress/support/commands.js`: `import 'cypress-file-upload'`

When I run my test, I get this error:
```
rror: Webpack Compilation Error
./cypress/support/commands.js
Module not found: Error: Can't resolve 'cypress-file-upload' in '/[myroot]/cypress/support'
resolve 'cypress-file-upload' in '[myroot]/cypress/support'
  Parsed request is a module
  using description file: [myroot]/package.json (relative path: ./cypress/support)
    Field 'browser' doesn't contain a valid alias configuration
    Looked for and couldn't find the file at the following paths:
```

I'm using yarn workspaces, so my `node_modules` directory is a higher up than my package.json where I'm requiring stuff.

**Comment by dragonfire1119** [comment]
> Can anyone help me figure out how to use the https://github.com/binti-family/cypress-file-upload/tree/cypress-v5-compat branch in my build?
> 
> 
> 
> What I did:
> 
> 1. Delete `cypress-file-upload` from node_modules directory.
> 
> 2. Delete `yarn.lock`.
> 
> 3. Add to `devDependencies` in `package.json`: `"cypress-file-upload": "binti-family/cypress-file-upload#cypress-v5-compat",
> 
> `
> 
> 4. `yarn install`
> 
> 5. Add to `cypress/support/commands.js`: `import 'cypress-file-upload'`
> 
> 
> 
> When I run my test, I get this error:
> 
> ```
> 
> rror: Webpack Compilation Error
> 
> ./cypress/support/commands.js
> 
> Module not found: Error: Can't resolve 'cypress-file-upload' in '/[myroot]/cypress/support'
> 
> resolve 'cypress-file-upload' in '[myroot]/cypress/support'
> 
>   Parsed request is a module
> 
>   using description file: [myroot]/package.json (relative path: ./cypress/support)
> 
>     Field 'browser' doesn't contain a valid alias configuration
> 
>     Looked for and couldn't find the file at the following paths:
> 
> ```
> 
> 
> 
> I'm using yarn workspaces, so my `node_modules` directory is a higher up than my package.json where I'm requiring stuff.
> 
> 

You have to go into the directory and yarn; yarn build. Yarn is not building the dist.

**Comment by KatieMFritz** [comment]
Thanks, @dragonfire1119 ! That did the trick.

**Comment by dragonfire1119** [comment]
> Thanks, @dragonfire1119 ! That did the trick.

You're welcome! Glad I could help! :)

**Comment by emilong** [comment]
@abramenal any chance you have time to check this out and either provide feedback or get a release going? really appreciate your work on this package and just hoping to unblock some folks :blush:

**Review by testtek — APPROVED** [review]

**Comment by testtek** [comment]
Shall we get this PR sooner to the release build? Blocking my test execution and refraining me from upgrade Cypress@5.0.0

**Comment by josephzidell** [comment]
Is there anything I can do to help move this along?

~ Another satisfied user of this important plugin

**Review by ThomasP1988 — APPROVED** [review]

File: lib/file/getFileBlobAsync.js
Comment (josephzidell) [inline lib/file/getFileBlobAsync.js:15]: This is an interesting way to check versions. However, this might fail once Cypress goes past version 9.

```js
'9.0.0' < '5.0.0' // => false
'10.0.0' < '5.0.0' // => true
```

Is there another way to check this?

**Comment by josephzidell** [comment]
Hey all! After speaking with @abramenal, it seems the best way forward is for me to become a maintainer on this project. Looking forward to working with you all!

This is my first time being an OSS maintainer, so please be gentle.

File: lib/file/getFileBlobAsync.js
Comment (josefsmrz) [reply lib/file/getFileBlobAsync.js:15]: [The Cypress documentation](https://docs.cypress.io/api/cypress-api/version.html) suggests to use the [semver](https://github.com/npm/node-semver) library.

Comment (emilong) [reply lib/file/getFileBlobAsync.js:15]: Nice catch!

How would you feel about including this module to do the checking? https://www.npmjs.com/package/compare-versions

Ideally I think it would be good to remove this check once Cypress has updated to remove the change that caused the issue, which hopefully comes in the 5.x.x line and well before 10.x.x.... :crossed_fingers:

Comment (emilong) [reply lib/file/getFileBlobAsync.js:15]: Thanks @josefsmrz - I'll use that instead :+1:

Comment (josephzidell) [reply lib/file/getFileBlobAsync.js:15]: @emilong As I'm thinking about it, I came to another way that might be safer than hinging on Cypress version. What if we check the return type? For example:

```js
let wrapBlob = blob => {
  // Cypress version 5 assigns a function with a compatibility warning
  // to blob.then, but that makes the Blob actually thenable. We have
  // to remove that to Promise.resolve not treat it as thenable.
  //
  // eslint-disable-next-line no-param-reassign
  delete blob.then;
  return blob instanceof Promise ? Cypress.Promise.resolve(blob) : blob;
};
```

This then has no dependencies. Would you be willing to try this?

**Comment by emilong** [comment]
Ok! Updated to use semver for comparison and I tried it locally on Cypress 4 and 5. Would appreciate anyone else double checking though!

File: lib/file/getFileBlobAsync.js
Comment (emilong) [reply lib/file/getFileBlobAsync.js]: Sure! I probably won't be able to get back to validating that this works until tonight/tomorrow though, but I like that way better!

**Review by abramenal — APPROVED** [review]

**Comment by josephzidell** [comment]
@abramenal Before this gets merged, I asked the author to try a different way that introduces no new dependencies and might also be safer. Hopefully we can wait 1 more day for that.

File: lib/file/getFileBlobAsync.js
Comment (emilong) [inline lib/file/getFileBlobAsync.js:9]: The eslint configuration forbids direct use of global `Promise` here and this approach appears to work on Cypress 4 in my tests, but we can do an exception if anyone finds an issue with this check.

**Comment by emilong** [comment]
I had a chance to update to use @josephzidell's approach and check it on Cypress 4 and 5 again.

File: lib/file/getFileBlobAsync.js
Comment (jennifer-shehane) [reply lib/file/getFileBlobAsync.js]: The `version` is available from the plugins `config` since 4.12.1 also https://on.cypress.io/writing-a-plugin#config

**Comment by josephzidell** [comment]
LGTM

**Comment by josephzidell** [comment]
@all-contributors add @emilong for code

**Comment by jojost1** [comment]
Will you also make a new release?

**Comment by josephzidell** [comment]
We're working on it 🍻

**Comment by abramenal** [comment]
@all-contributors add @josephzidell for maintenance

**Comment by abramenal** [comment]
Thanks everyone for working on this, `v4.1.0` is out 🎉🎉🎉

**Comment by josephzidell** [comment]
Unfortunately, Cypress 5 is still showing this error:

```
binaryStringToBlob() no longer returns a Promise. Update the use of binaryStringToBlob() to expect a returned Blob
```

**Comment by testtek** [comment]
Thank you !!! I can confirmed that it worked !! 👍

**Comment by martinsoenen** [comment]
It worked for me too !! Thanks a lot

**Comment by bbonevPIT** [comment]
I just pulled the latest code, I can confirm the file upload is working on the latest cypress version, thank you so much guys!

**Comment by JohnnyChiang** [comment]
> Unfortunately, Cypress 5 is still showing this error:
> 
> ```
> binaryStringToBlob() no longer returns a Promise. Update the use of binaryStringToBlob() to expect a returned Blob
> ```

Having the same error here

**Comment by Ksan8** [comment]
I'm also still getting the error that @josephzidell and @JohnnyChiang mentioned. Going to downgrade again until this is stable.

**Comment by emilong** [comment]
@Ksan8  @josephzidell @JohnnyChiang can you say more about your set up? E.g. I'm using cypress to drive chrome only and I didn't have any other plugins/extensions that hit this issue. Those are the first two things that come to mind as being possibilities?

**Comment by saschwarz** [comment]
@Ksan8 @josephzidell @JohnnyChiang take a look at #222 and #225 to see if that is your use case. Also, 4.1.1 has been released with a fix.

**Pull request merged**
